olap/src/aggregator/ceramic_patch.rs

```diff
@@ -0,0 +1,147 @@
+#[derive(Debug)]
+pub struct CeramicPatch {
+    signature: Signature,
+}
+
+impl CeramicPatch {
+    pub fn new_udwf() -> WindowUDF {
+        WindowUDF::new_from_impl(Self::new())
+    }
+    fn new() -> Self {
+        Self {
+            signature: Signature::new(
+                TypeSignature::Exact(vec![
+                    DataType::Binary,
+                    DataType::Binary,
+                    DataType::Utf8,
+                    DataType::Utf8,
+                ]),
+                Volatility::Immutable,
+            ),
+        }
+    }
+}
+
+impl WindowUDFImpl for CeramicPatch {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "ceramic_patch"
+    }
+
+    fn signature(&self) -> &datafusion::logical_expr::Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, _arg_types: &[DataType]) -> Result<DataType> {
+        Ok(DataType::Utf8)
+    }
+
+    fn partition_evaluator(&self) -> Result<Box<dyn PartitionEvaluator>> {
+        Ok(Box::new(CeramicPatchEvaluator))
+    }
+}
+
+#[derive(Debug)]
+struct CeramicPatchEvaluator;
+
+impl CeramicPatchEvaluator {
+    fn apply_patch(patch: &str, previous_state: &str) -> Result<String> {
+        let patch: Vec<json_patch::PatchOperation> = serde_json::from_str(patch)
+            .map_err(|err| exec_datafusion_err!("Error parsing patch: {err}"))?;
+        let mut new_state: serde_json::Value = serde_json::from_str(previous_state)
+            .map_err(|err| exec_datafusion_err!("Error parsing previous state: {err}"))?;
+        json_patch::patch(&mut new_state, &patch)
+            .map_err(|err| exec_datafusion_err!("Error applying JSON patch: {err}"))?;
+        serde_json::to_string(&new_state)
+            .map_err(|err| exec_datafusion_err!("Error JSON encoding: {err}"))
+    }
+}
+
+impl PartitionEvaluator for CeramicPatchEvaluator {
+    fn evaluate_all(&mut self, values: &[ArrayRef], num_rows: usize) -> Result<ArrayRef> {
+        let event_cids = as_binary_array(&values[0])?;
+        let previous_cids = as_binary_array(&values[1])?;
+        let previous_states = as_string_array(&values[2])?;
+        let patches = as_string_array(&values[3])?;
+        let mut new_states = StringBuilder::new();
+        for i in 0..num_rows {
+            if previous_cids.is_null(i) {
+                //Init event, patch value is the initial state
+                if patches.is_null(i) {
+                    // If we have an init event without data use an empty object as the initial
+                    // state. This feels like a leaky abstraction is this expected based on the
+                    // Ceramic spec?
+                    new_states.append_value("{}");
+                } else {
+                    new_states.append_value(patches.value(i));
+                }
+            } else if let Some(previous_state) = if !previous_states.is_null(i) {
```

Comment:
I found this `else if let Some() = if/else {}` nesting hard to follow at first.. but looking at it in my IDE it's alright.. and I'm not sure what I'd prefer so just mentioning it as an observation without any expectation

Reply:
Same not sure how to best simplify, i'll play around with a few thing